### PR TITLE
[MOD] Adjust colors

### PR DESCRIPTION
**Pull request by minnieoo**
The adjust colors mod allows a user to customize the colors of Kbin.

Features: 
- Customize upvote, downvote, and boost colors
- Sliders for sepia, hue, saturation, contrast, and brightness
- Option to add border to up/downvote buttons
- Reset button to reset values back to their defaults

image
image
firefox_Jcsc4eu2jI

### CODE REVIEW

**Comment by aclist** [comment]
This seems great! Simple but effective. I'll be taking a further look at it later.

File: helpers/manifest.json [hunk 1]
```diff
@@ -924,5 +924,108 @@
...
+        "key": "contrast",
+        "label": "Contrast",
+        "step": "0.1",
+        "min": "0",
+        "max": "10",
+        "show_value": true
+      },
+      {
+        "type": "color",
+        "initial": "#24B270",
```
Comment (aclist) [inline helpers/manifest.json:1008]: I was wondering where these default values come from--can we start by inheriting values from kbin, e.g., `var(--kbin-vote-text-color)`?

**Review by aclist — CHANGES_REQUESTED** [review]

File: helpers/manifest.json [hunk 2]
```diff
@@ -924,5 +924,108 @@
...
+        "key": "upvote",
+        "label": "Upvote color"
+      },
+      {
+        "type": "color",
+        "initial": "#DC2F3F",
```
Comment (aclist) [inline helpers/manifest.json:1014]: see above

File: helpers/manifest.json [hunk 3]
```diff
@@ -924,5 +924,108 @@
...
+        "key": "downvote",
+        "label": "Downvote color"
+      },
+      {
+        "type": "color",
+        "initial": "#FFFFFF",
```
Comment (aclist) [inline helpers/manifest.json:1020]: see  above

**Comment by minnieoo** [comment]
i wanted to do that, but if i am not mistaken, the initial value of the color type input has to be a hex value, so i didn't know how to go about it. like, i tried this:

```js
  let upvoteCol = settings.upvote;
  let downvoteCol = settings.downvote;
  let defaultUpvote = 'var(--voteup-color)';
  let defaultDownvote = 'var(--votedown-color)';

  .vote .active.vote__up button {
        color: ${upvoteCol ? upvoteCol : defaultUpvote};
        ${settings.border ? 'border: 2px solid ' + upvoteCol + ';' : ''}
    }
  .vote .active.vote__down button {
        color: ${downvoteCol ? downvoteCol : defaultDownvote};
        ${settings.border ? 'border: 2px solid ' + downvoteCol + ';' : ''}
    }
```

but then i realized it already is truthy since the default values of the input are and must be (?) set to a hex code. unless i am wrong? advice wanted lol

**Comment by aclist** [comment]
Oops, that is a good point, the value is being set right in the input field, so it has to be a hex value and can't be interpolated from a variable. In that case, try this:

`getComputedStyle(document.documentElement).getPropertyValue('--kbin-vote-text-color')`

You can parse that into a var and use that for the initial value.

**Comment by minnieoo** [comment]
omg i never heard of the computed styles thing, thats awesome.

>You can parse that into a var and use that for the initial value.

how can i use it for the initial value?

**Comment by aclist** [comment]
Sorry, my brain is not working :(

The manifest only supports hex values...got it...

I'll have to extend support for this on the KES side, I guess we could do one of the two below:

Regex approach: parse starting char for `-` or `#`
```
initial: --some-var
initial: #FFFFFF
```

JSON approach:
```
initial: {
 type: hex
 value: #FFFFFF
}
```

I think I'll probably take option one, seems like it can't ever be something other than a hash mark or hyphen.

**Comment by aclist** [comment]
OK, should be updated now, nice idea! :smiley_cat: 

You can use it like this: `"initial": "--some-var"`

**Comment by minnieoo** [comment]
YAYYYYY awesome ty ty, you're so fast :D

**Comment by minnieoo** [comment]
i put the var names in the initial field, it is showing black as the default now

image
image

**Comment by aclist** [comment]
I'm not seeing that as an available kbin var. The names are: 

`--kbin-upvoted-color` and `--kbin-downvoted-color`

**Comment by minnieoo** [comment]
oh lol idk where tf i got the other variables from haha. 

on a slightly related note, i'm wondering if i am testing my branch properly? what i do to test is just change the links in the kes.user.js to my branch within tampermonkey, is that the correct way to test? it does work. ex:

image

and if so, is there a way to make it update faster when i push changes to my branch? cause i will push a change, and be stuck waiting wondering if i did something wrong or if it is just being slow, and sometimes it is just being slow and the changes show up some minutes later, makes it really annoying to debug and test things. what i try to do to combat this is to change the version up and down, and switch this back and forth between always and never a few times, like turning something on and off again lol. 

image

i can see the changes in my branch so i dont understand why it sometimes takes forever for it to be received on tampermonkey. if there is more efficient ways to test, if i am doing something wrong, i would very much like to know

**Comment by aclist** [comment]
> on a slightly related note, i'm wondering if i am testing my branch properly? what i do to test is just change the links in the kes.user.js to my branch within tampermonkey, is that the correct way to test? it does work. ex:

"Formal method" is to run the `gen.sh` in the repository root. It's a bash shell script, should work on Mac as well in the terminal (probably) or Windows Subsystem for Linux (WSL), never tried either of those myself. It replaces all of the links with the names of the ones for the current slug/branch and spits out a new kes.user.js. Try it, it should contain the right URLs to your fork.

> and if so, is there a way to make it update faster when i push changes to my branch? cause i will push a change, and be stuck waiting wondering if i did something wrong or if it is just being slow, and sometimes it is just being slow and the changes show up some minutes later, makes it really annoying to debug and test things. what i try to do to combat this is to change the version up and down, and switch this back and forth between always and never a few times, like turning something on and off again lol.

I believe it has to do with the way these extensions handle caching. I've been thinking about a possible improvement, since it also might affect end-users if their extension silently updates the main KES script but doesn't update the new resources. I think going forward, we'll version all of the additional resources with the version change, so instead of `// @require URL...user.js`, it'll be `//  @require URL...-2.2.0-beta.12.user.js` When the resource names are different, extensions *should* always fetch them anew, since there is no cache. This'll get added into the `gen.sh` script when I have time, probably soon. The script also has to rename the mods and everything, so it's not just a matter of changing the URLs, but bumping everything.

> i can see the changes in my branch so i dont understand why it sometimes takes forever for it to be received on tampermonkey. if there is more efficient ways to test, if i am doing something wrong, i would very much like to know

Best thing I can recommend for now, and which I do, is to just blow away the current KES script from Tampermonkey by deleting it. This destroys all cache and will force it to download the current version of the remote resources each time.

Setting the externals to always is a bit risky, as it means any remote URL in any script you have installed is allowed to auto-update itself, so if the URL changes to contain malicious content, it's going to get pulled in automatically. This is set to never for security reasons. TBH, the auto-updating feature is itself a crutch/half-baked feature in these extensions. I talked about it a bit more [here](https://kbin.social/m/enhancement/t/251625/KES-2-1-0-Cornucopia-of-new-community-add-ons#entry-comment-1155867)

Anyway, I think it should be resolved in the long term once I switch to the version numbers in atomic files.

**Comment by minnieoo** [comment]
thanks! i'll look into the gen.sh thing, i am unfamiliar with that and bash shell. also, i can say for sure now that it is still showing up as black even though i corrected the variable names in the manifest. the reason i have white as the downvote initial value currently was to make 100% sure that my externals had updated

image

**Comment by aclist** [comment]
The var interpolation feature worked for me when I tested, will check it against your version and see what is different.

**Comment by aclist** [comment]
What them are you using btw?

**Comment by aclist** [comment]
```
$ getComputedStyle(document.documentElement).getPropertyValue('--kbin-upvoted-color')

→ "#0f5132" 
```
manifest.json
```
![a](https://github.com/aclist/kbin-kes/assets/92275929/303785aa-f64b-4cd5-b82f-6e0f2e58ccf4)

  {
    "name": "Label OP",
    "author": "shazbot",
    "version": "0.1.0",
    "label": "Add OP label",
    "login": false,
    "recurs": true,
    "desc": "Print OP label next to thread author",
    "entrypoint": "labelOp",
    "namespace": "labelcolors",
    "page": "threads",
    "fields": [
      {
        "type": "color",
        "initial": "--kbin-upvoted-color",
        "key": "fgcolor",
        "label": "Foreground color"
      },
      {
        "type": "color",
        "initial": "#111111",
        "key": "bgcolor",
        "label": "Background color"
      }
    ]
  },
```

a

**Comment by aclist** [comment]
Tested your script as well and it shows up correctly. Sounds like you still have an old manifest file clunking around in your cache.

**Comment by minnieoo** [comment]
>Tested your script as well and it shows up correctly. Sounds like you still have an old manifest file clunking around in your cache.

AHHHHHHHHHHHHHHH WHAT?? i even deleted the script and stuff like you suggested lmfao, why does this happen to meee 😭. it makes it SO hard to debug shit when half the time there is no issue, it's just being slow and dumb, and i break more things trying to fix something that isn't broken. thanks for letting me know lolll

another question, do we have a script to remove the random post sidebar on kbin yet? people keep asking for that, seems like a convenient script

**Comment by minnieoo** [comment]
it is just refusing to work for me on my end, so i am gonna assume from your tests that it's actually working and say it's good to go 👍🏼

**Comment by aclist** [comment]
Maybe you have multiple extensions or multiple versions of the script installed?

I mean, it may be a bug, I don't know technically how, but it could be, might as well isolate the root cause.

> another question, do we have a script to remove the random post sidebar on kbin yet? people keep asking for that, seems like a convenient script

Yes, check General > Hide sidebar elements. I updated it recently because people were asking for it.

**Comment by aclist** [comment]
Also, what theme are you using? Maybe it's a theme that sets upvotes to black?

**Comment by minnieoo** [comment]
i finally got it to work, i just recreated the script for the 100th time 🙃

>Yes, check General > Hide sidebar elements. 

awesome :D

**Comment by minnieoo** [comment]
nvm it is good to go 👍🏼

**Comment by minnieoo** [comment]
ok now it is good to go

**Comment by minnieoo** [comment]
wait hold on, need to fix one more thing

**Comment by minnieoo** [comment]
some issues. i cannot change the color. if i try to change it, accept, leave the panel and go back in, it's back to the initial value. there is definitely an issue with the fact that the default value is a string and not some sort of color code, 

image

i thought, well, i cant do this do this:

```js
   .vote .active.vote__down button {
        color: ${downvoteCol};
        ${settings.border ? `border: 2px solid var(${downvoteCol});` : ''}
    }
```

because after the user changes the color it will be a hex code, and var(#903DE) or whatever is not gonna work. 

so i tried this, where i instead make their choice into a var so i could simply wrap it all in var():

```js
    :root {
        chosen-upvote-color: ${upvoteCol}
        chosen-downvote-color: ${downvoteCol}
        chosen-boost-color: ${boostCol}
    }

    html {
        filter: sepia(${sepia}) hue-rotate(${hue}) brightness(${bright}) saturate(${saturate}) contrast(${contrast});
    }

    .vote .active.vote__up button {
        color: var(--chosen-upvote-color);
        ${settings.border ? `border: 2px solid var(--chosen-upvote-color);` : ''}
    }

    .vote .active.vote__down button {
        color: var(--chosen-downvote-color);
        ${settings.border ? `border: 2px solid var(--chosen-downvote-color);` : ''}
    }

    .entry footer menu > a.active, .entry footer menu > li button.active {
        color: var(--chosen-boost-color);
        text-decoration: none;
    }
```

but then i realized this issue, now the var says this when it is set to default, which is not valid:
image

i thought, okay ill even wrap that in var(), but wait, can't, because it will also be a hex code at some point and I have therefore looped back around to my top most issue in this message where it is not valid. any suggestions?

**Comment by aclist** [comment]
I think you should parse the incoming value in the same way KES does it:

```
let upvoteCol = getHex(settings.upvote);

function getHex (value) {
    const firstChar = Array.from(initial)[0];
    let realHex;
    if (firstChar === "-") {
        realHex = getComputedStyle(document.documentElement).getPropertyValue(initial);
    } else {
        realHex = value;
    }
    return realHex;
}
```

**Comment by aclist** [comment]
I also suggest using `safeGM("addStyle",myCss)`, it'll make inserting stylesheets easier than manipulating the DOM.

**Comment by aclist** [comment]
We can also look into making `getHex` a global function KES provides as a utility for mods that might need to parse kbin theme colors on the fly.

**Comment by minnieoo** [comment]
> safeGM("addStyle",myCss)

what do i do with that? replace this with it?

```js
adjustStyle = document.createElement('style');
document.head.appendChild(adjustStyle);
```

>I think you should parse the incoming value in the same way KES does it:

why am i getting the error `ReferenceError: initial is not defined`?

>We can also look into making getHex a global function KES provides as a utility for mods that might need to parse kbin theme colors on the fly.

sounds good 👍🏼

**Comment by aclist** [comment]
I had a typo here that said initial:

Before
```
function getHex (value) {
    const firstChar = Array.from(initial)[0];
```
After
```
function getHex (value) {
    const firstChar = Array.from(value)[0];
```

**Comment by minnieoo** [comment]
Sorry? I don't understand, that's the same thing I already had? unless i am having a ditzy moment

**Comment by aclist** [comment]
It's here, then:

Before
```
 if (firstChar === "-") {
        realHex = getComputedStyle(document.documentElement).getPropertyValue(initial);
```
After
```
 if (firstChar === "-") {
        realHex = getComputedStyle(document.documentElement).getPropertyValue(value);
```

**Comment by minnieoo** [comment]
ohhhh okay, and about this? `safeGM("addStyle",myCss)`

**Comment by aclist** [comment]
Here's how I would do it, it's up to you (didn't test it)

```
function adjustSite(toggle) {
    if (toggle) {
        adjustColors("custom");
    } else {
        adjustColors("revert")
    }
    function getHex (value) {
        const firstChar = Array.from(value)[0];
        let realHex;
        if (firstChar === "-") {
            realHex = getComputedStyle(document.documentElement).getPropertyValue(initial);
        } else {
            realHex = value;
        }
        return realHex;
    }
    function adjustColors(mode) {
        let settings = getModSettings('adjust');
        let sepia = `${settings.sepia * 10}%`;
        let hue = `${settings.hueRotate * 10}deg`;
        let bright = `${(settings.bright * 10) + 100}%`;
        let saturate = `${(settings.saturate * 10) + 100}%`;
        let contrast = `${(settings.contrast * 10) + 100}%`;
        let upvoteCol
        let downvoteCol
        let boostCol

        if (mode === "revert") {
            upvoteCol = 'initial'
            downvoteCol = 'initial'
            boostCol = 'initial'
        } else {
            upvoteCol = getHex(settings.upvote);
            downvoteCol = getHex(settings.downvote);
            boostCol = getHex(settings.boost);
        }

        const css = `
            html {
                filter: sepia(${sepia}) hue-rotate(${hue}) brightness(${bright}) saturate(${saturate}) contrast(${contrast});
            }
            .vote .active.vote__up button {
                color: ${upvoteCol};
                ${settings.border ? `border: 2px solid ${upvoteCol};` : ''}
            }
            .vote .active.vote__down button {
                color: ${downvoteCol};
                ${settings.border ? `border: 2px solid ${downvoteCol};` : ''}
            }
            .entry footer menu > a.active, .entry footer menu > li button.active {
                color: ${boostCol};
                text-decoration: none;
            }
        `;
        safeGM("addStyle",css)
    }
}
```

**Comment by minnieoo** [comment]
im trying it. also, something is reeeeally stumping me right now. the code is pretty much exactly what you suggested so I have no idea what the hell this is from but whenever i toggle the script on and off it keeps having this injected into the `<head>`. i have no idea why or where it is coming from, all scripts are off except this, ive cleared the cash, deleted and redid the tampermonkey script a few times completely.

https://github.com/aclist/kbin-kes/assets/137783738/fef143f6-0f5b-4670-bff8-544918413723

it keeps injecting this, which is what i had from an earlier test. pls help 🙃
```css

    html {
        filter: sepia(0%) hue-rotate(0deg) brightness(100%) saturate(100%) contrast(100%);
    }
    .vote .active.vote__up button {
        color: #0f5132;
        border: 2px solid #0f5132;
    }
    .vote .active.vote__down button {
        color: #FFFFFF;
        border: 2px solid #FFFFFF;
    }
    .entry footer menu > a.active, .entry footer menu > li button.active {
        color: #606060;
        text-decoration: none;
    }
 ```

**Comment by minnieoo** [comment]
ok i cleared the casce 4000 times and it worked now!?!?!?!? fuck this is frustrating. but the issue remains that i cannot change the color, it changes back

https://github.com/aclist/kbin-kes/assets/137783738/02106a89-cb3c-46a1-a197-6e3a6e655a39

**Comment by minnieoo** [comment]
also, i think the colors of the vars are being parsed wrong or something. they are much darker with the script on.
left is script off, right is on and the resulting hex code
image

**Comment by aclist** [comment]
Let's just use your way, it should be more reliable unless extra error handling is added. The thing about addStyle is there is no removeStyle equivalent, so it just keeps stacking stylesheets into the document head, with the newest one taking precedence provided the rules are explicitly set. It might produce unexpected results unless you handle cleanup of the sheet at position 0. As for the darkness, it's probably due to multiple sheets applying over each other and amplifying the opacity level.

**Comment by aclist** [comment]
Come to think of it, I should probably work up a removeStyle helper function to make adding/removing styles easier.

**Comment by minnieoo** [comment]
okay i reverted those changes, but the issue about the colors remains, they won't change. i get this error in console when i try to change the colors `Uncaught TypeError: json[it] is undefined`. they are also still, by default, very dark. no layering happening

script color
image

actual var color which i just plugged into here to show the difference
image

**Comment by aclist** [comment]
Alright, give this a try, it should be fixed. I tested each option back and forth and debugged everything, it's pretty much KES's fault due to some lacking features around this dynamic color stuff. Sorry I couldn't give you my undivided attention earlier, my focus was split between several things.

Basically, this was a handful of issues with KES happening all at once. For one, I didn't add sufficient support for the custom color fields with the theme variables. It was sort of working, but was lacking some core functionality I didn't realize in my haste to implement it.

Secondly, the catch reset button was not updated to account for the `getHex()` logic, so it wasn't resetting colors to the right value and just reverting to `#000000`.

Thirdly, there was some very funny business with the theme itself, which explained the dark color once I found the root cause. The updated `getHex()` should explain it:

```
function getHex (value) {
    const firstChar = Array.from(value)[0];
    let realHex;
    const theme = document.querySelector('.fixed-navbar');
    if (firstChar === "-") {
        realHex = getComputedStyle(theme).getPropertyValue(value);
    } else {
        realHex = value;
    }
    return realHex;
}
```

The element `.fixed-navbar` actually contains the theme definition, e.g., `theme--dark`, in its class name, and this element holds all of the color vars.

Without grabbing the contents of this element, we were just polling the global kbin vars from the document root, which are not the same as the currently selected theme, just some generic fallback ones.

Lastly, I updated the `addStyle` and `removeStyle` functions to support setting and removing stylesheets with a named ID, so that you can insert them and remove them from the head without leaving detritus behind. Calling logic is:

`safeGM("addStyle", cssbody, selector)`, where `selector` is some ID with anchor. You can see this in action in the filter script:

```
const sheetName = "#custom-kes-colors"
.
.
//later
safeGM("addStyle", customCSS, sheetName)
```

I might change this to drop the anchor so that it's more consistent with other JS functions for setting an element name, but it works.

Similarly, to remove a sheet you do the reverse:

```
safeGM("removeStyle", sheetName)
```

It's good practice to, when entering a function that applies a style, remove the named sheet first before applying a new one with change so that there are no collisions, and only one named sheet of that type exists:


```
function adjustColors (sheetName) {
    safeGM("removeStyle", sheetName)
// do stuff
// at the end:
    safeGM("addStyle", customCSS, sheetName)
}
```

I'll document this all later in the help files. You've really pushed me to make some handy utility functions/features with each of your scripts:

- Better support for number fields, min/max, step increments
- Step increments on range sliders
- Find and convert kbin theme colors to hex
- Insert and remove CSS styles dynamically

This is good, as we're abstracting the setup further so that mod authors can focus more on the design of mods rather than debugging and setting up complex functions to do stuff that should work out of the box.

The end result is a lot more flexible and easy to setup mods, without dealing with as much DOM manipulation. Especially with auto insertion and removal of stylesheets, now you can set up a whole boilerplate sheet with classes and just insert/remove it in one line, which should significantly cut down on the complexity of scripts themselves.

**Comment by aclist** [comment]
Note: I didn't try borders :fearful:

**Comment by minnieoo** [comment]
I am glad that my scripts can help progress the functionality of KES in some ways :D

I am trying it out now and getting this error:

image
And it is missing the color inputs for some reason:

image

I am pretty sure it is updated, I cleared the cache, waited, deleted and remade the script, tested on another browser, the usual stuff to get externals updated and it still shows this. Any idea why?

**Comment by aclist** [comment]
I'll take a look in a few minutes.

**Comment by aclist** [comment]
Try this version of KES, does it work? 

https://gist.github.com/aclist/bf91d9d5c8b12666bc6e55a0d3f43de6/raw/70862109d39216494aba046421f43292c0800cdf/kes.user.js

**Comment by minnieoo** [comment]
Still the same thing for me. did the usual again and tried the script in the link. also i only just now realized somehow that the requires and entry points are alphabetical lmfao

image

**Comment by aclist** [comment]
> Still the same thing for me. did the usual again and tried the script in the link.

It's working over here, those error messages are something I'm aware of but don't block functionality, they just need to be suppressed when getHex is called when toggle is off. Maybe your browser is more strict and is failing to draw the boxes when this happens? I'll try to resolve the errors with getHex first.

> also i only just now realized somehow that the requires and entry points are alphabetical lmfao

It's done automatically, hence why there is no need for the author to insert stuff here. Makes troubleshooting easier, too, because you find them in the list

**Comment by aclist** [comment]
Tried again and I'm still getting zero errors, actually. Are you sure you deleted everything :face_with_spiral_eyes: 

I stress tested it a lot and I don't think there are any problems. Have you tried in a clean Firefox prefix? (`firefox -P`)

**Comment by minnieoo** [comment]
i did the new profile thing in firefox just now, still the same thing, i really have no damn clue why :0 also wow, it's crazy seeing an uncluttered firefox lol

**Comment by aclist** [comment]
Yeah, ads agalore. I'm going to try merging the "working" version into KES proper so that we aren't carrying around all of these links, try pulling the latest version of testing and seeing if it works end to end. I'm not discounting the possibility that it's broken on your end, just failing to see why.

**Comment by minnieoo** [comment]
still the same thing, very strange. i just had my friend try this, he never had tampermonkey before, and it says the same thing for him, so i confirmed that it's not just a me issue

**Comment by aclist** [comment]
OK, so KES loads, you can open the menu, when you navigate to the adjust colors mod, it throws those errors in console? Can you click on the `getHex` and `adjustSite` names in the traceback and pinpoint on what line it's throwing the error?

**Comment by minnieoo** [comment]
The error is thrown the second KES loads in. When i click get hex I get this:

image

when i click adjustSite i get this:
image


which is within all this and more, it's rlly long but im putting it here to hopefully make it more clear where it is at? idk:
image
image

**Comment by aclist** [comment]
What's the output of `document.querySelector('.fixed-navbar')`?

**Comment by minnieoo** [comment]
null

**Comment by aclist** [comment]
Makes sense now. Logged in or not logged in? What theme did you set?

Seems the current method of parsing the applied kbin theme is not reliable, need to find a better way.

**Comment by minnieoo** [comment]
logged in. same thing when logged out. i am using kbin theme

**Comment by aclist** [comment]
Isn't that the bright white one?!

Alright, I'll investigate how we can check the theme more reliably, should be smooth sailing once that is figured out.

**Comment by aclist** [comment]
How about this? 

`document.querySelector('body').className`

**Comment by minnieoo** [comment]
image

>`document.querySelector('body').className`

output: `"theme--kbin KUP-injected kup-show-article-preview vsc-initialized"`

**Comment by aclist** [comment]
OK, kind of bizarre that other mods inject classnames all the way up to the body. I guess we can parse the class as an array and get index 0, then. I pushed a fix on testing, try now.

**Comment by minnieoo** [comment]
still no :C 
image

this is the console on opera if it matters
image

**Comment by aclist** [comment]
Should be fixed. I forgot that the whole element contains the style definitions, don't even need to parse the theme name.

**Comment by minnieoo** [comment]
YEEEEEESSSSS IT WORKS NOW!! what a beautiful sight. thanks for that!! 🥳🥳

**Comment by aclist** [comment]
Sorry it took so long. :face_exhaling: 

Good job on the bug smashing.

**Comment by minnieoo** [comment]
i didnt do anything, you're the one who did a good job hahaha. its no problem, i don't think either of us expected the issue to run as deep as it did lol

**Comment by aclist** [comment]
Let's test this again tomrrow and then merge it.

**Comment by minnieoo** [comment]
sounds good. also, i was wondering if it would be a good idea to have an option to show visual examples (images or gifs?) of what the scripts do, if the mod authors choose to add it? i know you don't want to crowd up the description area, so it could be toggle-able in the accessibility panel perhaps, or maybe on the top where the transparency button is and so on. i know a lot of scripts I didn't understand what it did until i turned it on, despite their descriptions, so i think it could be good for general accessibility reasons and just more clear as well. like for example:

image
image

**Comment by aclist** [comment]
I'm generally against something overengineered/fancy not for aesthetic reasons, but because it opens up a can of worms of technical issues that you have to support. For example, cons I can think of: 

- Slower loading time from fetching these images
- Need to standardize and normalize the size and quality of all images
- Constrained size may not fit all details if image needs a before/after comparison
- Necessarily makes the settings difficult to reach

Then there's the issue of, in the next update a mod that blurs/unblurs NSFW images is going in. Imagine putting the before/after pic in the KES preview, hahah.

Luckily, there is a solution to this: the `link` field and the `images` directory. Currently there isn't a single mod using this. Originally we put it in for the help files to your emoticons mod. But we could repurpose this as a link to a preview image.

I think this is good, as it puts this feature to use and keeps the content out of the menu, and the remote images can be any size/layout/dimensions. Finally, it takes away the performance problem and the NSFW problem because seeing the previews is on-demand if the user chooses to open it.

**Comment by minnieoo** [comment]
>Then there's the issue of, in the next update a mod that blurs/unblurs NSFW images is going in. Imagine putting the before/after pic in the KES preview, hahah.

I would expect that this is self-explanatory enough to not warrant a visual example

>Luckily, there is a solution to this: the link field and the images directory. Currently there isn't a single mod using this. Originally we put it in for the help files to your emoticons mod. But we could repurpose this as a link to a preview image.

I totally forgot about the link field lol. Maybe some people would want to use that for external documentation, I am not sure, so it being repurposed is up to you. maybe the link and preview image link can be different fields for diff purposes, idk. but yea, do whatever you'd like, i just thought of it and thought i'd bring it up but it isn't overly necessary so i am not bothered about whether it is or isnt implemented :D

**Comment by aclist** [comment]
> I would expect that this is self-explanatory enough to not warrant a visual example

Never assume anything, that's the rule of software design :D

> I totally forgot about the link field lol. Maybe some people would want to use that for external documentation, I am not sure, so it being repurposed is up to you. maybe the link and preview image link can be different fields for diff purposes, idk. but yea, do whatever you'd like, i just thought of it and thought i'd bring it up but it isn't overly necessary so i am not bothered about whether it is or isnt implemented :D

I'm in favor of it, do you want to take some screengrabs and slap them into the `/images` directory? I'm saying that since we have the link field and it's not being used by anybody and probably won't be, we basically have a readymade solution to this problem. Can just rename it to "Image preview" and stuff the link in there. I think this is a good suggestion because people do want to see a preview, but if we go down the rabbit hole of inline previews, it adds a lot of fiddly issues and inevitably is going to look wacky on e.g. mobile.

**Comment by aclist** [comment]
For that matter, if someone wants to put documentation in there (with its own images), they still can, that's the beauty of this solution. I think forcing images on the user is an anti-pattern, whereas making them dynamically available should you choose to open them solves the bandwidth, mobile, customization issues in one swoop.

By the way, it's not quite finished yet, but I added a GUI manifest generator to the repo under `manifest_gen`. You can open the file `sc.html` in your browser and it provides an interface for populating custom fields and generating the JSON from that. 

It's not done yet. Might need your help with making the design a bit more intuitive to use.

**Comment by minnieoo** [comment]
>Never assume anything, that's the rule of software design :D

if anything, i'd expect something like this to be done in that case, where the example isnt actually nsfw, but i still get your point, we can do this anyway if we are going the preview links route:

image

>do you want to take some screengrabs and slap them into the /images directory? 

yea sure i can do that. do we want them in `/images` or perhaps a separate folder like `/screenshots` or `/previews`? there might be quite a few of them, not sure if it matters if they are mixed up with the current example images and the kibby images but thought i'd ask anyway. also what is the field called? links? would i be doing something like `"link": "/images/mod-preview.png"` on the mods i take the sc's for?

>For that matter, if someone wants to put documentation in there (with its own images), they still can, that's the beauty of this solution

How would you add both preview images and external doc link?

>You can open the file sc.html in your browser and it provides an interface for populating custom fields and generating the JSON from that.

this looks great! for my next mods I am gonna try it out more. i think for the next one i'm gonna do the tag paste functionality mod without the modal.
firefox_KIF5s9d8SA

>Might need your help with making the design a bit more intuitive to use.

Bold of you to assume I am any better in CSS than i am in JS 🤣😅

**Comment by aclist** [comment]
> if anything, i'd expect something like this to be done in that case, where the example isnt actually nsfw, but i still get your point, we can do this anyway if we are going the preview links route:

Hahaha, that's a good piont, not sure why I didn't think of that.

> yea sure i can do that. do we want them in /images or perhaps a separate folder like /screenshots or /previews? there might be quite a few of them, not sure if it matters if they are mixed up with the current example images and the kibby images but thought i'd ask anyway. 

Previews sounds good

> also what is the field called? links? would i be doing something like "link": "/images/mod-preview.png" on the mods i take the sc's for?

It's described [here](https://aclist.github.io/kes/kes.html#_json_manifest), see the example `manifest.json` with the `link` and `link_label` fields

> How would you add both preview images and external doc link?

Well, you create an external doc and you embed an image in it, of course ((smiling innocently))

> this looks great! for my next mods I am gonna try it out more. i think for the next one i'm gonna do the tag paste functionality mod without the modal.

I should have the actual JSON output finished soon, currently it handles the barebones but not the custom fields. The image preview I'll probably leave for a bit until I have time to do it, it not essential.

> Bold of you to assume I am any better in CSS than i am in JS

:face_in_clouds:

**Comment by aclist** [comment]
By the way, I probably wouldn't unnecessarily highlight stuff in the screenshot, it's liable to be misunderstood that the mod adds some border or highlight. Also, try to keep the size of the images smallish.

**Comment by aclist** [comment]
I have the updated omnibar working on testing, care to take a look? I believe everything below is working:

- Found via **General > Subscriptions omnibar**. Set a meta key used to invoke the omnibar
- On toggle on: if no cached mags, fetch them and store them (happens seamlessly in the background)
- If logged out, show a selection of the top mags from the magazines index
- If logged in, show the user's subscribed mags
- Once mags are cached, they won't be fetched again unless the mod is disabled/enabled again (faster load)
- On a given page, you can invoke the meta key an indefinite amount of times to open/close the omnibar
- You can also click anywhere outside the omnibar to close it
- Supports up/down arrow keys and mouse hover to select a mag
- If using the mouse and then the arrow keys, will remember the last hovered position
- Search-as-you-type to filter the results
- Hit enter or click to jump to the mag (typical fast keyboard usage would be: invoke meta key => type to search => arrows => enter)
- When toggling the mod to off, will reload once to wipe event listeners

Problems!!

Theme colors are mostly perfect except for the blasted purple night theme and the default kbin theme. Think you can find a better color alternative?

**Comment by minnieoo** [comment]
>I have the updated omnibar working on testing, care to take a look? 

where is it? 😭😭 am i looking in the wrong spot?

image

**Comment by aclist** [comment]
You're on an old manifest again, make sure to delete KES entirely from TM and reinstall

**Comment by aclist** [comment]
You should be on `2.2.0-beta.67`

**Comment by minnieoo** [comment]
i was on the wrong branch 😭 i was on testing omni rather than testing, my bad. it appears to be working very good :D the purple theme version looks fine to me btw, whats wrong with it?

image

**Comment by aclist** [comment]
Yeah, sorry, I merged most of this into testing.

Oo, I was actually talking about the kbin light theme, the theme from hell. Currently the hover color is this bizarre grape tone.

What's going on with your scrollbar though?

**Comment by minnieoo** [comment]
>What's going on with your scrollbar though?

idk lol thats how it looks on firefox, the scrollbar is light colored, works fine tho.

https://github.com/aclist/kbin-kes/assets/137783738/414cd59e-9f57-442d-ae2a-25851373701f

i have a theme on tho that makes it dark with this
```css
    body {
        scrollbar-color: #d4d4d43b #7878787a;
    }
```
image

>Oo, I was actually talking about the kbin light theme, the theme from hell.

oh i see it now lol. using the root color vars it's hard to find one that looks good on light theme without messing up the other themes, it truly is the theme from hell 😭

**Comment by aclist** [comment]
> idk lol thats how it looks on firefox, the scrollbar is light colored, works fine tho.

Ah, I meant that the scrollbar shouldn't be visible when you constrain the list to only a few items.

**Comment by minnieoo** [comment]
>Ah, I meant that the scrollbar shouldn't be visible when you constrain the list to only a few items.

ohhh, yea no its still there for me no matter what,
image

**Comment by aclist** [comment]
Hmm, I don't have that. Better test it on your vanilla Firefox prefix :wink:

**Comment by minnieoo** [comment]
same thing 
image

**Comment by aclist** [comment]
We need a CSS guru on this project :face_exhaling:

**Comment by minnieoo** [comment]
lord knows that aint me. im sitting here proud of tiny stuff like making random elements 5% cuter lol. doing this is so fun for me. i love the new noti panel btw, its so much faster :0

image

could it be that the scroll bar is still visible because the elements that are hidden by the search are still taking up space in the container? so the container is not resizing to fit the remaining visible elements?

**Comment by aclist** [comment]
> lord knows that aint me. im sitting here proud of tiny stuff like making random elements 5% cuter lol.

This is the best metric

> i love the new noti panel btw, its so much faster :0

Thanks! What's the mockup with the fat read/purge buttons? The right side of the notipanel header is reserved for forward/back buttons (pagination), if you have more than one page of notis.

> could it be that the scroll bar is still visible because the elements that are hidden by the search are still taking up space in the container? 

Yes, that's precisely correct. I was able to reproduce the dangling scrollbar thing on Chrome, working up a fix. Changing the height property to `fit-content` seems to do the trick.

**Comment by minnieoo** [comment]
>Thanks! What's the mockup with the fat read/purge buttons? The right side of the notipanel header is reserved for forward/back buttons (pagination), if you have more than one page of notis.

yup it's still there if i have another page. the buttons are wide bc i prefer it that way lol, i liked it better than them off to the left and small, personal preference. it isnt a mockup it's a custom stylus. :
image

>Yes, that's precisely correct. I was able to reproduce the dangling scrollbar thing on Chrome, working up a fix. Changing the height property to fit-content seems to do the trick.

oh awesome :D

**Comment by aclist** [comment]
Ok, it *should* be fixed now. There is also a special bar on mobile at the top of the screen that you can tap to open the omnibar instead of using the keyboard.

**Comment by minnieoo** [comment]
>Ok, it should be fixed now.

working good

**Comment by aclist** [comment]
> working good

Thank you. Still need to tweak some things.

I am merging this PR, I think it is working correctly now. We can integration test it further inside KES.

**Comment by aclist** [comment]
I just noticed something while testing some other mods that use the color picker. Seems like if the user has set a custom color and it's saved in settings, the color picker still reverts to showing the "default" color instead of their chosen color. I think you mentioned this before.

**Comment by aclist** [comment]
Fixed this in 703fc5d

**Pull request merged**